mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java

```diff
@@ -0,0 +1,187 @@
+    Schema schema = Schema.parseJson(config.schema);
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    for (Schema.Field field : schema.getFields()) {
+      builder.set(field.getName(), convertValue(bsonObject.get(field.getName()).toString(), field));
+    }
+    emitter.emit(builder.build());
+  }
+
+  private Object convertValue(String input, Schema.Field field) throws IOException {
+    switch (field.getSchema().getType()) {
+      case NULL:
+        return null;
+      case BOOLEAN:
+        return Boolean.valueOf(input);
+      case INT:
+        return Integer.valueOf(input);
+      case LONG:
+        return Long.valueOf(input);
+      case FLOAT:
+        return Float.valueOf(input);
+      case DOUBLE:
+        return Double.valueOf(input);
+      case BYTES:
+        return input.getBytes();
+      case STRING:
+      case ENUM:
+        return input;
+      case ARRAY:
+        return GSON.fromJson(input, List.class);
+      case MAP:
+        return GSON.fromJson(input, Map.class);
+      case UNION:
+        if (field.getSchema().isNullableSimple()) {
+          return convertValue(input, Schema.Field.of(field.getName(), field.getSchema().getNonNullable()));
+        } else {
+          return null;
+        }
+    }
+    throw new IOException(String.format("Unsupported schema : %s for field '%s'",
+                                        field.getSchema(), field.getName()));
+  }
+
+  public static class MongoDBConfig extends PluginConfig {
+
+    @Name(Properties.CONNECTION_STRING)
+    @Description("MongoDB Connection String (see http://docs.mongodb.org/manual/reference/connection-string); " +
+      "Example: 'mongodb://localhost:27017/analytics.users'.")
+    private String connectionString;
+
+    @Name(Properties.AUTH_CONNECTION_STRING)
+    @Nullable
+    @Description("Auxiliary MongoDB connection string to authenticate against when constructing splits.")
+    private String authConnectionString;
+
+    @Name(Properties.SCHEMA)
+    @Description("The schema for the data as it will be formatted in CDAP. Sample schema: {\n" +
+      "    \"type\": \"record\",\n" +
+      "    \"name\": \"schemaBody\",\n" +
+      "    \"fields\": [\n" +
+      "        {\n" +
+      "            \"name\": \"name\",\n" +
+      "            \"type\": \"string\"\n" +
+      "        },\n" +
+      "        {\n" +
+      "            \"name\": \"age\",\n" +
+      "            \"type\": \"int\"\n" +
+      "        }" +
+      "    ]\n" +
+      "}")
+    private String schema;
+
+    @Name(Properties.INPUT_QUERY)
+    @Description("Optionally filter the input collection with a query. This query must be represented in JSON " +
+      "format, and use the MongoDB extended JSON format to represent non-native JSON data types.")
+    @Nullable
+    private String inputQuery;
+
+
+    @Name(Properties.INPUT_FIELDS)
+    @Nullable
+    @Description("A projection document limiting the fields that appear in each document.")
```

Comment:
not sure what a 'projection document' is, is this mongo terminology? Let's add to the description what happens if this is not given.